### PR TITLE
Update TSTORE and TLOAD opcode values to match Eip1153 updates

### PR DESCRIPTION
**Pull request by Demuirgos**
## Changes

- updates TSTORE/TLOAD opcodes

#### What types of changes does your code introduce?

- [ ] Bugfix (a non-breaking change that fixes an issue)
- [x] New feature (a non-breaking change that adds functionality)
- [ ] Breaking change (a change that causes existing functionality not to work as expected)
- [ ] Optimization
- [ ] Refactoring
- [ ] Documentation update
- [ ] Build-related changes
- [ ] Other: _Description_

#### Requires testing

- [x] Yes
- [ ] No

#### If yes, did you write tests?

- [x] Yes
- [ ] No

#### Notes on testing

_Optional. Remove if not applicable._

## Remarks

_Optional. Remove if not applicable._

### CODE REVIEW

File: src/Nethermind/Nethermind.Evm/VirtualMachine.cs [hunk 1]
```diff
@@ -1999,28 +1956,88 @@ private CallResult ExecuteCall<TTracingInstructions>(EvmState vmState, byte[]? p
...
+                case Instruction.BEGINSUB | Instruction.TLOAD:
                     {
-                        if (!spec.SubroutinesEnabled) goto InvalidInstruction;
+                        if (spec.TransientStorageEnabled)
+                        {
 
-                        // why do we even need the cost of it?
-                        if (!UpdateGas(GasCostOf.Base, ref gasAvailable)) goto OutOfGas;
+                            Metrics.TloadOpcode++;
+                            var gasCost = GasCostOf.TLoad;
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Evm/VirtualMachine.cs:1965]: remove local variable

File: src/Nethermind/Nethermind.Evm/VirtualMachine.cs [hunk 2]
```diff
@@ -1999,28 +1956,88 @@ private CallResult ExecuteCall<TTracingInstructions>(EvmState vmState, byte[]? p
...
+
+                            // why do we even need the cost of it?
+                            if (!UpdateGas(GasCostOf.Base, ref gasAvailable)) goto OutOfGas;
+
+                            goto InvalidSubroutineEntry;
+                        }
 
-                        goto InvalidSubroutineEntry;
                     }
                 case Instruction.RETURNSUB:
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Evm/VirtualMachine.cs:1993]: or TSTORE

File: src/Nethermind/Nethermind.Evm/VirtualMachine.cs [hunk 3]
```diff
@@ -1999,28 +1956,88 @@ private CallResult ExecuteCall<TTracingInstructions>(EvmState vmState, byte[]? p
...
-                        if (!UpdateGas(GasCostOf.Low, ref gasAvailable)) goto OutOfGas;
+                            if (vmState.IsStatic) goto StaticCallViolation;
 
-                        if (vmState.ReturnStackHead == 0)
-                        {
-                            goto InvalidSubroutineReturn;
+                            long gasCost = GasCostOf.TStore;
+                            if (!UpdateGas(gasCost, ref gasAvailable)) goto OutOfGas;
+
+                            stack.PopUInt256(out UInt256 storageIndex);
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Evm/VirtualMachine.cs:2004]: same here

**Review by LukaszRozmej — CHANGES_REQUESTED** [review]

File: src/Nethermind/Nethermind.Evm/VirtualMachine.cs (hunk 2 above)
Comment (Demuirgos) [reply src/Nethermind/Nethermind.Evm/VirtualMachine.cs]: wont this line only be reached if SubroutinesEip is enabled and TransientStorageOpcodes are disabled?

**Review by LukaszRozmej — APPROVED** [review]

**Review by benaadams — APPROVED** [review]

**Review by MarekM25 — APPROVED** [review]

**Review by smartprogrammer93 — APPROVED** [review]

File: src/Nethermind/Nethermind.Evm/VirtualMachine.cs [hunk 4]
```diff
@@ -1994,28 +1951,75 @@ private CallResult ExecuteCall<TTracingInstructions>(EvmState vmState, byte[]? p
...
-                        gasAvailable -= GasCostOf.Base;
+                            stack.PopUInt256(out result);
+                            storageCell = new(env.ExecutingAccount, result);
+
+                            byte[] value = _state.GetTransientState(in storageCell);
+                            stack.PushBytes(value);
+
+                            if (typeof(TTracingStorage) == typeof(IsTracing))
+                            {
+                                _txTracer.LoadOperationTransientStorage(storageCell.Address, result, value);
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Evm/VirtualMachine.cs:1969]: Potential issue for tracing as gas check is moved later. @benaadams

File: src/Nethermind/Nethermind.Evm/VirtualMachine.cs [hunk 5]
```diff
@@ -1994,28 +1951,75 @@ private CallResult ExecuteCall<TTracingInstructions>(EvmState vmState, byte[]? p
...
+                            stack.PopUInt256(out result);
+                            storageCell = new(env.ExecutingAccount, result);
+                            bytes = stack.PopWord256();
+
+                            _state.SetTransientState(in storageCell, !bytes.IsZero() ? bytes.ToArray() : BytesZero32);
+
+                            if (typeof(TTracingStorage) == typeof(IsTracing))
+                            {
+                                byte[] currentValue = _state.GetTransientState(in storageCell);
+                                _txTracer.SetOperationTransientStorage(storageCell.Address, result, bytes, currentValue);
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Evm/VirtualMachine.cs:2004]: Potential issue for tracing as gas check is moved later. @benaadams

File: src/Nethermind/Nethermind.Evm/VirtualMachine.cs [hunk 6]
```diff
@@ -1994,28 +1951,75 @@ private CallResult ExecuteCall<TTracingInstructions>(EvmState vmState, byte[]? p
...
-                        // why do we even need the cost of it?
-                        gasAvailable -= GasCostOf.Base;
+                            stack.PopUInt256(out result);
+                            storageCell = new(env.ExecutingAccount, result);
+
+                            byte[] value = _state.GetTransientState(in storageCell);
+                            stack.PushBytes(value);
+
+                            if (typeof(TTracingStorage) == typeof(IsTracing))
+                            {
```
Comment (benaadams) [inline src/Nethermind/Nethermind.Evm/VirtualMachine.cs:1968]:
```suggestion
                            {
                                if (gasAvailable < 0) goto OutOfGas;
```

File: src/Nethermind/Nethermind.Evm/VirtualMachine.cs [hunk 7]
```diff
@@ -1994,28 +1951,75 @@ private CallResult ExecuteCall<TTracingInstructions>(EvmState vmState, byte[]? p
...
+                            gasAvailable -= GasCostOf.TStore;
+
+                            stack.PopUInt256(out result);
+                            storageCell = new(env.ExecutingAccount, result);
+                            bytes = stack.PopWord256();
+
+                            _state.SetTransientState(in storageCell, !bytes.IsZero() ? bytes.ToArray() : BytesZero32);
+
+                            if (typeof(TTracingStorage) == typeof(IsTracing))
+                            {
```
Comment (benaadams) [inline src/Nethermind/Nethermind.Evm/VirtualMachine.cs:2002]:
```suggestion
                            {
                                if (gasAvailable < 0) goto OutOfGas;
```

**Review by benaadams — COMMENTED** [review]
As @LukaszRozmej points out I'd add a gas check at start of trace so the trace doesn't change

**Pull request merged**